Pull Request: Adding HiRA integration into PEFT library

hqsiswiliam opened the pull request:
### Feature request


This request proposes integrating **HiRA** (Hadamard High-Rank Adaptation) as described in the ICLR 2025 oral paper (https://openreview.net/pdf?id=TwJrTz9cRS) (https://iclr.cc/virtual/2025/oral/31839) and implemented in the [hqsiswiliam/hira](https://github.com/hqsiswiliam/hira) repository into the core PEFT library. This will enable users to apply HiRA through the familiar `get_peft_model` API and benefit from its high-rank updates without adding any inference overhead.  

### Motivation

## General Motivation

PEFT methods like LoRA achieve parameter-efficient fine-tuning by injecting low-rank updates into pre-trained weights. While effective, purely low-rank adaptation can struggle to capture complex patterns in large language models.

### 1. Expressiveness grows with the rank

Empirically, increasing the LoRA rank in LLM training yields better downstream performance:

![LoRA performance vs. rank](https://github.com/hqsiswiliam/hira/blob/master/Figures/lora_perf.png?raw=true)  
*Higher LoRA rank correlates with improved task accuracy.*  

### 2. HiRA: Hadamard high-rank updates without extra parameters

HiRA sidesteps the expressiveness constraint by computing a Hadamard-enhanced update:

$$
\Delta W = W_0 \odot (A B)
$$

![HiRA update formula](https://github.com/hqsiswiliam/hira/blob/master/Figures/HiRA%20formula.png?raw=true)  
HiRA uses the Hadamard product to inject high-rank structure into the frozen weight matrix $W_0$ via low-rank matrix $A$ and $B$.

### 3. Singular-value patterns

After training, HiRA exhibits a rich singular-value pattern, akin to full-rank fine-tuning (FFT), indicating its ability to model complex transformations without the expensive computational overhead:

![Singular-value pattern comparison](https://github.com/hqsiswiliam/hira/blob/master/Figures/singular-value-pattern.png?raw=true)  
*HiRA’s singular-value distribution closely mirrors that of FFT.*  

### 4. Performance gains

Across commonsense reasoning benchmarks, HiRA outperforms LoRA and other PEFT baselines:

![Commonsense reasoning performance](https://github.com/hqsiswiliam/hira/blob/master/Figures/commonsense_reasoning.png?raw=true)  
*HiRA delivers notable accuracy improvements over baseline adapters.*  

### 5. No extra parameter or compute cost

Despite its high-rank behaviour, HiRA introduces no additional trainable parameters compared to LoRA:

![Resource comparison: LoRA vs. HiRA](https://github.com/hqsiswiliam/hira/blob/master/Figures/res.png?raw=true)  
*HiRA matches LoRA’s GRAM usage and training hours.*  

### 6. Complementary with LoRA (HiLoRA)

Combining HiRA and LoRA into a hybrid “HiLoRA” setup yields even stronger results than either method alone:

![HiLoRA concept](https://github.com/hqsiswiliam/hira/blob/master/Figures/HiLoRA.png?raw=true)  
![HiLoRA performance gains](https://github.com/hqsiswiliam/hira/blob/master/Figures/hilora_performance.png?raw=true)  
*HiLoRA leverages both low-rank and Hadamard high-rank updates for better expressiveness.*  

---

By integrating HiRA into PEFT, users gain richer adaptation capability without sacrificing the parameter efficiency and usability that PEFT provides.  


### Your contribution

We would be pleased to submit a pull request to integrate HiRA class implementation into the PEFT framework. We welcome any suggestions for alternative integration approaches and appreciate any guidance on best practices.

hqsiswiliam committed:
- initial commit for hira adapter

hqsiswiliam committed:
- initial commit for hira adapter

hqsiswiliam committed:
- This initial modification of HiRA's config

hqsiswiliam committed:
- This initial modification of HiRA's config

hqsiswiliam committed:
- update HiRA Model

hqsiswiliam committed:
- update HiRA Model

hqsiswiliam committed:
- update HiRA Layer

hqsiswiliam committed:
- update HiRA Layer

hqsiswiliam committed:
- update HiRA Layer partially

hqsiswiliam committed:
- update HiRA Layer partially

hqsiswiliam committed:
- update HiRA Layer partially (Embedding Layer)

hqsiswiliam committed:
- update HiRA Layer partially (Embedding Layer)

hqsiswiliam committed:
- update HiRA Layer partially (ConvNd Layer)

hqsiswiliam committed:
- update HiRA Layer partially (ConvNd Layer)

hqsiswiliam committed:
- update HiRA Layer partially (ConvNd Layer)

hqsiswiliam committed:
- update HiRA Layer partially (ConvNd Layer)

hqsiswiliam committed:
- update HiRA Layer partially (Conv1/2/3d Layer)

hqsiswiliam committed:
- update HiRA Layer partially (Conv1/2/3d Layer)

hqsiswiliam committed:
- update HiRA Layer partially (MultiheadAttention)

hqsiswiliam committed:
- update HiRA Layer partially (MultiheadAttention)

hqsiswiliam committed:
- remove HiRA Layer partially (MultiheadAttention)

hqsiswiliam committed:
- remove HiRA Layer partially (MultiheadAttention)

hqsiswiliam committed:
- update HiRA `layer`, `model`, and `config`

hqsiswiliam committed:
- update HiRA `layer`, `model`, and `config`

hqsiswiliam committed:
- add bnb implementation and __init__.py

hqsiswiliam committed:
- add bnb implementation and __init__.py

hqsiswiliam committed:
- add HiRA's Linear8bitLt implementation

hqsiswiliam committed:
- add HiRA's Linear8bitLt implementation

hqsiswiliam committed:
- update HiRA's layer comment

hqsiswiliam committed:
- update HiRA's layer comment

hqsiswiliam committed:
- add HiRA's Linear4bit

hqsiswiliam committed:
- add HiRA's Linear4bit

hqsiswiliam committed:
- complete HiRA's Linear4bit

hqsiswiliam committed:
- complete HiRA's Linear4bit

hqsiswiliam committed:
- add test_hira

hqsiswiliam committed:
- add test_hira

hqsiswiliam committed:
- HiRA: updates to peft init, tuners, types, and GPU tests

hqsiswiliam committed:
- HiRA: updates to peft init, tuners, types, and GPU tests

hqsiswiliam committed:
Merge remote-tracking branch 'upstream/main'

hqsiswiliam committed:
- HiRA: updates to HiRA layer, and HiRA testing

hqsiswiliam committed:
- HiRA: updates to HiRA layer, and HiRA testing

hqsiswiliam committed:
- HiRA: formatting hira

hqsiswiliam committed:
- HiRA: formatting hira

hqsiswiliam committed:
- HiRA: formatting hira

hqsiswiliam committed:
- HiRA: formatting hira

hqsiswiliam committed:
- HiRA: add document

hqsiswiliam committed:
- HiRA: add document

hqsiswiliam committed:
- apply merge

BenjaminBossan reviewed (changes requested):
Thanks for this PR to add HiRA to PEFT. The method looks promising and the provided code is already quite mature.

When I started reading the paper, I was at first reminded of [FedPara](https://huggingface.co/papers/2108.06098), aka LoHa, which is already integrated into PEFT, as that method also relies on the Hadamard product. However, IIUC, the two methods are still distinct: HiRA basically corresponds to LoRA, but instead of adding dW, we multiply it. In that way, it is much closer to LoRA than to LoHa. Still, I wanted to flag this, as I'm not sure you are aware (your paper doesn't seem to be reference FedPara).

At the moment, I haven't done a full in-depth review, but I think that makes more sense once we have completed the next step.

I noticed that you have formatted some unrelated files in `method_comparison`, could you please undo those changes? Usually, when you run `make style`, that directory should not be included.

I think a good next step is to add HiRA to the testing matrix we have in PEFT. For now, let's add some entries similar to the ones you can find here:

https://github.com/huggingface/peft/blob/92d65cafa51c829484ad3d95cf71d09de57ff066/tests/test_custom_models.py#L70-L72

Since you also support embedding and conv layers, please make sure to include examples with those layers as well (basically, copy the relevant examples from LoRA and adjust them).

Then, please run `pytest tests/test_custom_models.py -k "hira and not shira" -v` and see if those tests pass. Once we get there, we can discuss the best next steps.

BenjaminBossan commented:
@hqsiswiliam Do you still plan on working on this PR?

hqsiswiliam commented:
> @hqsiswiliam Do you still plan on working on this PR?

Hi, [BenjaminBossan](https://github.com/BenjaminBossan). Thanks for checking in! I’ll continue working on this PR over the next few days.

hqsiswiliam committed:
- undo method_comparison formatting

hqsiswiliam committed:
- update HiRA cases in test_custom_models.py

hqsiswiliam committed:
- update HiRA cases in test_custom_models.py

hqsiswiliam committed:
- update HiRA Embedding cases in test_custom_models.py

hqsiswiliam committed:
- update HiRA Embedding cases in test_custom_models.py

hqsiswiliam committed:
- add suggested changes from BenjaminBossan

hqsiswiliam committed:
- add suggested changes from BenjaminBossan

hqsiswiliam committed:
- add fix for ConvND in HiRA Layer

hqsiswiliam committed:
- add fix for ConvND in HiRA Layer

hqsiswiliam committed:
- update test for HiRA

hqsiswiliam committed:
- update test for HiRA

hqsiswiliam committed:
- update copyright notice.

hqsiswiliam committed:
- update copyright notice.

hqsiswiliam committed:
- update copyright notice.

hqsiswiliam committed:
- fix HiRA's config and model for updated fork sync.

hqsiswiliam committed:
Merge remote-tracking branch 'origin/main'

# Conflicts:
#	src/peft/tuners/hira/config.py
#	src/peft/tuners/hira/model.py
#	tests/test_custom_models.py

hqsiswiliam commented:
Hi, sorry for the long delay. I have updated the code and synced it with the latest changes in the repository. Could you please reopen this PR so I can push the updates? Thanks a lot for your time and help.

BenjaminBossan commented:
Thanks, done @hqsiswiliam

hqsiswiliam committed:
- remove module mapping duplication

hqsiswiliam committed:
- remove HiRARuntimeConfig

hqsiswiliam committed:
- update copyright information
- update r's description

BenjaminBossan commented on docs/source/package_reference/hira.md:
We also need an entry in the toctree for this to show up in the docs.

BenjaminBossan commented on src/peft/tuners/hira/model.py:
```diff
@@ -0,0 +1,761 @@
+        self.active_adapter = adapter_name
+
+    @contextmanager
+    def _enable_peft_forward_hooks(self, *args, **kwargs):
```
I'd say let's remove `_enable_peft_forward_hooks` from HiRA for now. It's not trivial to support, we should get the basics first and then later we can think about whether we want to add it or not.

BenjaminBossan commented on src/peft/tuners/hira/model.py:
```diff
@@ -0,0 +1,761 @@
+
+        return combination_type, new_r, new_target_modules
+
+    def add_weighted_adapter(
```
Let's also remove this for now and focus on the basics. As is, it also wouldn't work because other methods called in here are missing.

BenjaminBossan commented on src/peft/tuners/hira/config.py:
```diff
@@ -0,0 +1,230 @@
+
+
+@dataclass
+class HiRAConfig(PeftConfig):
```
Let's rename this to `HiraConfig`.

BenjaminBossan commented on src/peft/tuners/hira/model.py:
```diff
@@ -0,0 +1,761 @@
+        except AttributeError:
+            pass
+
+        quant_methods = ["gptq", "aqlm", "awq"]
```
Only bnb is supported, right? Let's remove the code dealing with other quant methods.

BenjaminBossan commented on src/peft/tuners/hira/layer.py:
```diff
@@ -0,0 +1,937 @@
+            msg = "Cannot pass `adapter_names` when there are merged adapters, please call `unmerge_adapter` first."
+            raise ValueError(msg)
+
+    def _mixed_batch_forward(
```
This method and its calls below can be removed if we drop `_enable_peft_forward_hooks`.

BenjaminBossan commented on src/peft/tuners/hira/layer.py:
```diff
@@ -0,0 +1,937 @@
+            weight_A = weight_A.float()
+            weight_B = weight_B.float()
+        output_tensor = transpose((weight_B @ weight_A), self.fan_in_fan_out)
+        assert self.get_base_layer().weight.shape == output_tensor.shape
```
Let's not use any `assert`s in the code (except for tests). I think this can be removed, if you want to keep it, raise a proper error.

BenjaminBossan reviewed (changes requested):
Thanks for your work on adding HiRA to PEFT. I focused in this review on the core implementation for now. A lot of my comments stem from the fact that this PR is based on the LoRA code with changes made to accommodate HiRA, but some of the LoRA code doesn't make sense here. Please check my comments on this.

I also thought about suggesting to add HiRA as a LoRA variant, similar to how DoRA is currently implemented. I think this would be a possibility to save a lot of code. Conceptually, however, I think HiRA is sufficiently different from LoRA that I wouldn't consider it a LoRA variant. LMK what you think about it.

Please merge with/rebase on the latest main branch and, once you finish your changes, please call `make style`.

hqsiswiliam commented:
Hi @BenjaminBossan, thanks a lot for the detailed review. I’m working through the comments and will update the PR accordingly. Thanks again for the helpful feedback!

hqsiswiliam committed:
- keeping only test_manual_hira_linear_equivalence

hqsiswiliam committed:
- remove unload_and_optionally_merge_module in hira model

hqsiswiliam committed:
- remove unsupported features

hqsiswiliam committed:
- remove unnecessary assert

hqsiswiliam committed:
- remove _mixed_batch_forward

hqsiswiliam committed:
- remove _check_forward_args and _mixed_batch_forward

hqsiswiliam committed:
- remove  quant_methods = ["gptq", "aqlm", "awq"]

hqsiswiliam committed:
- remove _cache_store and _cache_pop

hqsiswiliam committed:
- remove
        kwargs = locals().copy()
        del kwargs["self"]

hqsiswiliam committed:
- remove _register_custom_module

hqsiswiliam committed:
- remove layer_replication

hqsiswiliam committed:
- remove add_weighted_adapter, _check_add_weighted_adapter
- rename HiRAConfig-> HiraConfig
- rename HiRAModel -> HiraModel
- remove _enable_peft_forward_hooks
- rename HiRALayer -> HiraLayer

hqsiswiliam committed:
- remove init_hira_weights

hqsiswiliam committed:
Simplify HiRA model to match latest LoRA structure

hqsiswiliam committed:
Merge pull request #1 from hqsiswiliam/codex/update-hira/model.py-based-on-lora/model.py

Simplify HiRA model to match latest LoRA structure

hqsiswiliam committed:
- apply newest fork to main branch

hqsiswiliam committed:
Align HiRA forward hook setup

hqsiswiliam committed:
Merge pull request #3 from hqsiswiliam/codex/update-hira/model.py-with-latest-lora/model.py

Align HiRA model with LoRA model

BenjaminBossan commented:
@hqsiswiliam Once you're done with your changes, please ping me so that I know the PR is ready for review.

hqsiswiliam committed:
- remove layer_replication
- remove _custom_modules
- fix NaN error when unmerging

hqsiswiliam committed:
- fix too large learning rate when testing HiRA to result NaN
- fix unmerge behaviours, only replace 0 with tiny numbers for numeric stable.

hqsiswiliam committed:
- set init_weight=True by default avoid randomness when not set.

hqsiswiliam committed:
Merge branch 'huggingface:main' into main

hqsiswiliam committed:
- do `make style` after all changes

BenjaminBossan reviewed (changes requested):
Thanks a lot for all the updates. I have checked them and did another review of the PR. There is still a bit of work to do, but the rest should be relatively simple.

There is now a merge conflict, which is fortunately easy to resolve. Could you please merge with/rebase on the latest main and fix it?

Also, as we're nearing completion, let's add an example to the `examples/` folder. You can add a new one, perhaps something from the paper, or copy an existing one and modify it to use HiRA.

Regarding testing, the existing tests are quite comprehensive but there are still uncovered cases. I added some comments about this, but there is more: HiRA tests should be added for decoder models, encoder-decoder models, etc. This is quite straightforward. Check for instance how this was added for [another PEFT method](https://github.com/huggingface/peft/pull/2678/files#diff-f691a80c5fbe336138af718016fd86299b7c0265b1f51c73577c709dcdfe3545) and take the same approach.

hqsiswiliam committed:
- remove EETQ code

hqsiswiliam committed:
- update _toctree.yml, adding HiRA to sections.

hqsiswiliam committed:
- rename rank_pattern for naming consistency.

hqsiswiliam committed:
- remove test_hira.py
- modify config.__class__ to config_cls
- remove _check_merge_allowed in hira model.

hqsiswiliam committed:
- resolving an issue when mulit-HiRA to one module.
- adding HiRA cases to `MULTIPLE_ACTIVE_ADAPTERS_TEST_CASES`.

hqsiswiliam committed:
- add encoder-decoder test cases.

hqsiswiliam committed:
- add example for hira

BenjaminBossan commented:
@hqsiswiliam There is a merge conflict, could you please take care? Also, once you're finished with your changes, please ping me for another review.

hqsiswiliam committed:
Merge remote-tracking branch 'upstream/main'

# Resolve Conflicts:
-	tests/test_custom_models.py

BenjaminBossan reviewed (changes requested):
There is a small error in the docs. Also, could you please run `make style` to ensure that the linter is happy?

hqsiswiliam committed:
- update hira.md

hqsiswiliam committed:
Merge remote-tracking branch 'upstream/main'

hqsiswiliam committed:
- run `make style`

BenjaminBossan commented on docs/source/package_reference/hira.md:
```diff
@@ -0,0 +1,90 @@
+
+Default layer replacement for HiRA adapters:
+
+[[autodoc]] tuners.hira.dispatch.dispatch_default
```
Let's remove this, it should resolve the docs builder issue.

BenjaminBossan reviewed (changes requested):
Thanks for the updates. There are still a few tests that are failing, could you please check?

hqsiswiliam committed:
- remove BitsAndBytes Integration and Dispatch Handler in hira.md

hqsiswiliam committed:
Merge upstream/main into main

BenjaminBossan reviewed (changes requested):
Thanks for updating the PR. Generally, not much is missing, but I found a few smaller issues, please check. A few HiRA tests are still failing, could you please check and resolve these errors?

On top of this, let's also add tests to `test_config.py`, `test_decoder_models.py`, `test_feature_extraction_models.py`, and `test_seq_classifier.py`. Moreover, I would strongly suggest to add an experiment to the [MetaMathQA benchmark](https://github.com/huggingface/peft/tree/main/method_comparison#how-to-add-new-experiments). This has the advantage that we can run an experiment to check that training works as expected and is more or less aligned with the expectations from the paper.

hqsiswiliam committed:
REFAC Remove unused code and simplify imports in model components

BenjaminBossan reviewed (changes requested):
Thanks for the latest updates.

> Thank you for your review. I have updated the code accordingly. I will also look into integrating an experiment on the MetaMathQA benchmark following the provided guidelines.

Sounds good. Please ping me once that is done, + the missing unit tests I mentioned in my last comment.

hqsiswiliam committed:
- change newline to LF

hqsiswiliam committed:
- resolve issues

hqsiswiliam committed:
Merge upstream/main

hqsiswiliam commented:
Hi @BenjaminBossan, sorry for the delay that led to this being auto-closed.

I've completed the remaining items from your Mar 21 review:
- Added HiRA entries to test_config.py, test_decoder_models.py, test_feature_extraction_models.py, and test_seq_classifier.py
- Re-ran all HiRA tests locally
- Merged the latest main

Could you please reopen this PR for another review?

BenjaminBossan commented on src/peft/tuners/hira/model.py:
```diff
@@ -0,0 +1,239 @@
+        parent,
+        current_key,
+        *,
+        parameter_name: Optional[str] = None,
```
Is this argument needed?

hqsiswiliam replied:
removed

BenjaminBossan commented on src/peft/tuners/hira/layer.py:
```diff
@@ -0,0 +1,812 @@
+        self,
+        adapter_name,
+        r,
+        hira_dropout,
```
We recently made a change in PEFT regarding the `update_layer` usage. Instead of passing the individual parameters like `hira_dropout` here, we pass the full `HiraConfig` and `update_layer` can pick the arguments it needs. This is to simplify the interface. The PR can be found here: #3047. For HiRA, this means that the `update_layer` and `__init__` methods need to be updated (both normal and bnb layers). And in `HiraModel`, the call site needs to be updated accordingly.

Check the linked PR, it should not be too much work. LMK if you have a question.

hqsiswiliam replied:
Thanks for the pointer! Refactored `update_layer` and `__init__` (for both normal and bnb layers) to accept `HiraConfig` instead of individual parameters, and updated the call site in `HiraModel` accordingly.

BenjaminBossan reviewed (changes requested):
Thanks for the last updates, there is not much that's still required.

> I will also look into integrating an experiment on the MetaMathQA benchmark following the provided guidelines.

This is still missing, did you have time to check? Don't hesitate to ask me if anything is unclear.

hqsiswiliam committed:
Add HiRA MetaMathQA experiment results

hqsiswiliam committed:
Use explicit default 0.1 for warmup_step_ratio

hqsiswiliam committed:
Fix get_optimizer_and_scheduler signature to single line

hqsiswiliam committed:
Merge remote-tracking branch 'upstream/main'

hqsiswiliam committed:
Remove unused parameter_name argument from _create_and_replace

hqsiswiliam committed:
Pass HiraConfig to update_layer and __init__ instead of individual parameters

hqsiswiliam committed:
Update HiRA MetaMathQA results (test acc=0.5307)

hqsiswiliam commented:
Thanks for the review! I've made the following updates:
- Both comments addressed.
- Added a MetaMathQA experiment at `method_comparison/MetaMathQA/experiments/hira/llama-3.2-3B-rank32-lr4e-3/`.
- Merged with the latest upstream main.

BenjaminBossan commented on method_comparison/MetaMathQA/utils.py:
```diff
@@ -87,6 +87,7 @@ class TrainConfig:
         optimizer_type: The name of a torch optimizer (e.g. AdamW) or a PEFT method ("lora+", "lora-fa")
         optimizer_kwargs: The optimizer keyword arguments (lr etc.)
         lr_scheduler: The learning rate scheduler (currently only None or 'cosine' are supported)
+        warmup_step_ratio: Fraction of total steps used for LR warmup (only relevant when lr_scheduler='cosine'), defaults to WARMUP_STEP_RATIO (0.1)
```
Is this change necessary? I'd rather keep the benchmarking code fix for this PR. We can discuss adding the warmup ratio in a separate PR.

hqsiswiliam replied:
Reverted, utils.py and run.py are back to upstream. Will submit the warmup ratio change as a separate PR.

BenjaminBossan commented on method_comparison/MetaMathQA/results/hira--llama-3.2-3B-rank32-lr4e-3.json:
Results look nice, thanks for running. But let's remove the result file, we will run the experiment on our own machine to ensure that the different results are comparable.

hqsiswiliam replied:
Certainly, the result was removed.

BenjaminBossan commented on method_comparison/MetaMathQA/experiments/hira/llama-3.2-3B-rank32-lr4e-3/training_params.json:
```diff
@@ -0,0 +1,8 @@
+    "weight_decay": 0.0
+  },
+  "warmup_step_ratio": 0.05,
+  "use_gc": true
```
Let's not use gradient checkpointing, as all the other experiments also run without it.

hqsiswiliam replied:
Removed gradient checkpointing from the experiment config. I originally enabled it because my local 16GB GPU would run out of GRAM.

BenjaminBossan reviewed (changes requested):
Thanks for your updates, the layer initialization now conforms with the rest of PEFT and the experiment looks promising. There are a few issues there which I flagged, otherwise the PR is ready.

hqsiswiliam committed:
Rename HiRA MetaMathQA experiment to lr5.7e-3, remove warmup_step_ratio and use_gc overrides

BenjaminBossan commented:
Please ping me when the PR is ready for review.

hqsiswiliam committed:
Use lr=4.5e-3 for HiRA MetaMathQA experiment

hqsiswiliam committed:
Merge remote-tracking branch 'upstream/main'

hqsiswiliam commented:
Hi, @BenjaminBossan, all review comments addressed:
- Reverted utils.py/run.py changes (will submit warmup_step_ratio as a separate PR).
- Removed gradient checkpointing from the experiment config.
- Tuned the learning rate for better stability with the default warmup.
- Also synced with the latest upstream `main` to avoid merge conflicts.

BenjaminBossan commented:
Thanks for your latest changes. I ran the benchmark on my machine and got a test accuracy of 0.498 at 22.5 GB max memory. A final run will follow later, once the PR is merged.

The CI is still complaining because of style checks. Could you please run `make style` and push the changes? You may require ruff v0.12.12 to match the CI.

hqsiswiliam committed:
Run make style

hqsiswiliam commented:
> Thanks for your latest changes. I ran the benchmark on my machine and got a test accuracy of 0.498 at 22.5 GB max memory. A final run will follow later, once the PR is merged.
> 
> The CI is still complaining because of style checks. Could you please run `make style` and push the changes? You may require ruff v0.12.12 to match the CI.

Done, ran `make style` with ruff 0.12.12 and pushed the changes. The CI should be green now.

BenjaminBossan commented:
@hqsiswiliam It seems that not all required changes were applied to ruff. Did you ensure that it sees the `pyproject.toml` that comes with PEFT? You can check the [log](https://github.com/huggingface/peft/actions/runs/25403213002/job/74589208267?pr=2668), the issue is that the order if imports is incorrect.

hqsiswiliam committed:
Fix import ordering in tuners/__init__.py

BenjaminBossan reviewed (changes requested):
The linter is now good, thanks for fixing. Some tests are, however, failing. Could you please check? Once finished, please run

`pytest tests/test_custom_models.py -k hira -v`

locally to ensure that the tests pass before pushing.

hqsiswiliam committed:
Fix AttributeError in get_delta_weight for Conv and Embedding layers

hqsiswiliam committed:
Relax merge tolerance for HiRA on Conv layers (CPU float accumulation)

hqsiswiliam commented:
> The linter is now good, thanks for fixing. Some tests are, however, failing. Could you please check? Once finished, please run
> 
> `pytest tests/test_custom_models.py -k hira -v`
> 
> locally to ensure that the tests pass before pushing.

Fixed. Verified locally with `pytest tests/test_custom_models.py -k hira -v` and CI is green now.

BenjaminBossan reviewed (approved):
Thanks for the last push. Everything looks great from my side.

hqsiswiliam commented:
Thanks a lot for the review and help through this. Really appreciate your patience and guidance.
